src/Raven.Server/Rachis/Leader.cs

```diff
@@ -789,12 +789,28 @@ private void EmptyQueue()
 
         internal static ConvertResultAction GetConvertResult(CommandBase cmd)
         {
+            ConvertResultAction action;
             switch (cmd)
             {
                 case AddOrUpdateCompareExchangeBatchCommand batchCmpExchangeCommand:
-                    return new ConvertResultAction(batchCmpExchangeCommand.ContextToWriteResult, CompareExchangeCommandBase.ConvertResult);
+                    action = batchCmpExchangeCommand.ConvertResultAction;
+                    if (action != null)
+                        return action;
+
+                    action = new ConvertResultAction(batchCmpExchangeCommand.ContextToWriteResult, CompareExchangeCommandBase.ConvertResult);
+                    Interlocked.CompareExchange(ref batchCmpExchangeCommand.ConvertResultAction, action, null);
```

Comment:
Can this be concurrent? If so, what do you do for the *loser* of this?

Reply:
the next line :)

Reply:
I don't follow. You are trying to _publish_ this value, but you may lose here.
What happens there?

Reply:
I don't care if I win or lose, all callers must get the _same_ instance